### PR TITLE
Move character case table declarations from src to detail

### PR DESCRIPTION
**Pull request by davidwendt**
Moves the `special_case_mapping` struct and the `get_special_case_hash_index` functions from `src` to detail header `char_tables.hpp` so they can be used from strings-udf functions. Also moved are the set of utilities for dissecting a flag value from the character type table.

### CODE REVIEW

File: cpp/include/cudf/strings/detail/char_tables.hpp [hunk 1]
```diff
@@ -62,6 +89,17 @@ const character_cases_table_type* get_character_cases_table();
  */
 const struct special_case_mapping* get_special_case_mapping_table();
 
+/**
+ * @brief Get the special mapping table index for a given code-point.
+ *
+ * @see cpp/src/strings/char_types/char_cases.h
+ */
```
Comment (davidwendt) [inline cpp/include/cudf/strings/detail/char_tables.hpp:96]: Moved from `char_cases.h`

**Review by karthikeyann — APPROVED** [review]
LGTM 👍

File: cpp/include/cudf/strings/detail/char_tables.hpp [hunk 2]
```diff
@@ -62,6 +89,17 @@ const character_cases_table_type* get_character_cases_table();
...
+constexpr uint16_t get_special_case_hash_index(uint32_t code_point)
+{
+  constexpr uint16_t special_case_prime = 499;  // computed from generate_special_mapping_hash_table
```
Comment (ttnghia) [inline cpp/include/cudf/strings/detail/char_tables.hpp:99]: It is recommended to use UPPER_CASE for `constexpr` values if possible.

**Review by ttnghia — APPROVED** [review]

File: cpp/include/cudf/strings/detail/char_tables.hpp (hunk 2 above)
Comment (davidwendt) [reply cpp/include/cudf/strings/detail/char_tables.hpp:99]: I don't see the recommendation documented. It appears this assertion is not true for most (>1000) cases in libcudf.

Comment (ttnghia) [reply cpp/include/cudf/strings/detail/char_tables.hpp:99]: It's not officially recommended in cudf. I vaguely remember that I've read that convention somewhere else and tried to follow it closely. Anway, it's just an unofficial recommendation.

**Comment by davidwendt** [comment]
@gpucibot merge

**Pull request merged**